openedx/core/djangoapps/xblock/olx_bundle/__init__.py

```diff
@@ -0,0 +1,238 @@
+            'linked_bundle/unit/unit1/definition.xml'
+        ]
+        """
+        cache_key = ("olx_file_list", )
+        cached = self.cache.get(cache_key)
+        if cached is not None:
+            return cached
+
+        bundle_files = get_bundle_files(self.bundle_uuid, use_draft=self.draft_name)
+        olx_files = [path for path in bundle_files if path.endswith("/definition.xml")]
+
+        self.cache.set(cache_key, olx_files)
+        return olx_files
+
+    def get_usage_definition(self, usage_id):
+        """
+        Get the BundleDefinitionLocator for the specified usage_id string,
+        in the most efficient way possible.
+
+        Returns None if the usage_id doesn't exist in this bundle.
+        """
+        # First, check if there is a definition file that matches the usage ID.
+        # This is fastest because it never involves parsing OLX file contents.
+        suffix = "/{}/definition.xml".format(usage_id)
+        for olx_path in self.get_olx_files():
+            if olx_path.endswith(suffix):  # Ends with so we don't need to know the type
+                block_type = olx_path.split('/')[-3]
+                return BundleDefinitionLocator(
+                    bundle_uuid=self.bundle_uuid,
+                    block_type=block_type,
+                    olx_path=olx_path,
+                )
+        # If that didn't fine the file, the usage ID is not a "top-level" block
+        # in this content library.
+        # Instead, one of the OLX files in this bundle contains something like:
+        # <xblock-include usage="custom-name" definition="html/acdef1234"/>
+        child_usages_map = self.get_usages_map()
+        return child_usages_map.get(usage_id, None)
+
+    def get_usages_map(self):
+        """
+        Get the map of usage_id strings -> BundleDefinitionLocator
+        """
+        _tld, usage_aliases_map = self._scan_bundle_olx()
+        return usage_aliases_map
+
+    def get_top_level_definitions(self):
+        """
+        Get the set of BundleDefinitionLocator items in this bundle that
+        have no parent. (Excludes blocks from linked bundles.)
+        """
+        top_level_definitions, _ua = self._scan_bundle_olx()
+        return top_level_definitions
+
+    def _scan_bundle_olx(self):
+        """
+        Helper method to scan bundles that follow certain conventions and return
+        both a list of "top level" blocks and a map of usage IDs to definitions.
+
+        ** In the future this method can/should be optimized away by performing
+           this computation on commit and writing the result into an olx.map
+           file within the bundle itself **
+
+        ** This method follows the semantics of course and library bundles, but
+           other types of bundles may not follow the same patterns. **
+
+        A course or library bundle consists of OLX files that MUST be named
+            {block_type}/{definition_id}/definition.xml
```

Comment:
This does not match this format: https://github.com/open-craft/blockstore/pull/27/files#diff-b97d1edf108582cb60ade48369663178R49

Reply:
Yeah, thinking on that evolved in subsequent conversations with @ormsbee . Do you have any strong preference for one or the other?

Reply:
I myself like the one we discussed in the referenced PR more. :) Was the subsequenct conversation on GitHub somewhere?

Reply:
Not on GitHub, but it is the bundle format used in https://github.com/open-craft/openedx-blockstore-relay/pull/9 which you reviewed and approved. I also forwarded you an email with some of the discussion.

Reply:
Okay, I read the email and now I understand this better. A couple of points:

1. If I remember correct a `block_id` is not necessarily unique in a course. It needs to be paired with a `block_type`. I don't know how big a problem that is and/or whether we even need to support that? @ormsbee?
2. Why does `get_usage_definition()` only take a `usage_id` and then iterate over the list of files? Any specific reason why we can't pass the `block_type` as well so that there is a one-to-one mapping?

Reply:
Those questions are related. In modulestore, `block_id` is not unique and must always be paired with `block_type`. In the new bundle format I have made the `id` part unique, so it is both sufficient and necessary for identifying any XBlock.

> Why does get_usage_definition() only take a usage_id and then iterate over the list of files? Any specific reason why we can't pass the block_type as well so that there is a one-to-one mapping?

In part to help enforce that the block_type is not part of the ID.

In part because it needs to check the list of files anyways to confirm that the file name of `block_type/block_id/definition.xml` actually exists, and I think that the difference between iterating over the list to find a specific suffix and checking for the presence of a known item in the list is negligible.

You are right that this could be optimized by also passing in the block type, but I don't think it would make a measurable difference. Perhaps in extremely large bundles with tens of thousands of blocks.

Reply:
> In modulestore, block_id is not unique and must always be paired with block_type. In the new bundle format I have made the id part unique, so it is both sufficient and necessary for identifying any XBlock.

But that implies that when we migrate over courses, this bundle format will not work for them?

Reply:
The way I interpreted the [initial proposal](https://github.com/open-craft/blockstore/pull/27/files#diff-b97d1edf108582cb60ade48369663178R49) is the following:

1. For every course there is going to be a main bundle.
2. The definition of a migrated block in a course with usage_key `block-v1:edX+DemoX+Demo_Course+type@html+block@030e35c4756a4ddc8d40b95fbbfff4d4` will exist at `type@html+block@030e35c4756a4ddc8d40b95fbbfff4d4/definition.xml` in that main bundle.
3. If a course adds an XBlock from a library created in blockstore, it can use the block-v2 key to refer to it.

> In part to help enforce that the block_type is not part of the ID.

The only code that needs to deal with the string value of the ID is the runtime when doing definition lookups. Isn't 
`type@html+block@030e35c4756a4ddc8d40b95fbbfff4d4` or `html/030e35c4756a4ddc8d40b95fbbfff4d4` as good an ID as `030e35c4756a4ddc8d40b95fbbfff4d4`?

Reply:
> But that implies that when we migrate over courses, this bundle format will not work for them?

First, it's not a problem for 99% of Open edX content because the `block_id` values are Mongo ObjectIds are so are already unique, regardless of the type.

Second, the key/IDs will be changing when importing to blockstore anyways (block-v1 -> block-v2 at least), so it's not a problem to enforce `block_id` uniqueness during import by rewriting any non-unique IDs to become unique as needed.

> 2. The definition of a migrated block in a course with usage_key block-v1:edX+DemoX+Demo_Course+type@html+block@030e35c4756a4ddc8d40b95fbbfff4d4 will exist at type@html+block@030e35c4756a4ddc8d40b95fbbfff4d4/definition.xml in that main bundle.

I don't remember what, if anything, we discussed about this. I also don't have any strong opinions on it.

One can just as easily say that `block-v1:edX+DemoX+Demo_Course+type@html+block@030e35c4756a4ddc8d40b95fbbfff4d4 ` should map to `type@html+block@html030e35c4756a4ddc8d40b95fbbfff4d4/definition.xml` (combining the type and ID together to form the new ID) which is a perfectly reversible transform while still allowing new block types to have only their ID unique.

However, what I actually implemented in [openedx-blockstore-relay](https://github.com/open-craft/openedx-blockstore-relay) is that `block-v1:edX+DemoX+Demo_Course+type@html+block@030e35c4756a4ddc8d40b95fbbfff4d4 ` maps to `html/030e35c4756a4ddc8d40b95fbbfff4d4/definition.xml` in the bundle. I haven't yet defined how this would look as an opaque key (no block-v2 for courses yet), but if it's similar to what I have proposed for libraries it would be something like `block-v2:edX+DemoX+Demo_Course:html:030e35c4756a4ddc8d40b95fbbfff4d4`

> The only code that needs to deal with the string value of the ID is the runtime when doing definition lookups.  Isn't
type@html+block@030e35c4756a4ddc8d40b95fbbfff4d4 or html/030e35c4756a4ddc8d40b95fbbfff4d4 as good an ID as 030e35c4756a4ddc8d40b95fbbfff4d4?

Yes, and if you'll look through this PR, you'll see that all the opaque keys do include both type and ID. I think it may only be this one internal method, and/or maybe a couple other internal methods, that are based on ID only without block type. I'm doing that because:

(1) In the past I've found the (block_type + block_id) uniqueness confusing and a source of bugs so I want to start off for now by enforcing block_id uniqueness.

(2) It's easy to change back to having only (block_type + block_id) is unique at any time going forward, but it's impossible to change back to only block_id being unique if we don't start with that.

(3) Either way, it's an implementation detail at the moment - the only user visible change is that the system enforces unique IDs at import time. That may result in a tiny number of courses that need to (manually or automatically) have some IDs transformed to be unique when importing.

BTW this UI in Studio and some functionality like the `jump_to_id` links today assume that IDs are unique, even though it's not enforced. So in some ways I consider the lack of enforcement in Studio today to be a bug:
![studio-screenshot](https://user-images.githubusercontent.com/945577/62828912-ceccbc00-bba6-11e9-8d07-54b9111e75a4.png)

Reply:
> Second, the key/IDs will be changing when importing to blockstore anyways (block-v1 -> block-v2 at least), so it's not a problem to enforce block_id uniqueness during import by rewriting any non-unique IDs to become unique as needed.

Just to make sure we are talking about the same thing, I was referring to the migration of live course content to Blockstore from Modulestore. In that case I think we will want to be able to keep using the same keys?

Thanks for all the additional details btw.

@ormsbee Your thoughts/preferences?

Reply:
> I was referring to the migration of live course content to Blockstore from Modulestore

which part of the [courseware migration plan](https://github.com/open-craft/blockstore/pull/33/files) is that? My understanding is there is no point in time where a live course is suddenly changed from Modulestore to Blockstore. Between runs, we'll be changing some courses to be _authored_ in Blockstore (and then published to modulestore and viewed there), and eventually to be authored + viewed from Blockstore.

Reply:
I think @symbolist might be thinking of folks who have git as a primary authoring mechanism and do a one way push of data into Studio/LMS for publishing. In that case, they would have to make changes to avoid conflicts if they have such ID conflicts. But I think it's fine to think of that as an edge case where we flag with errors and ask people to adjust their content.